### PR TITLE
[determinism] Add d9m-unimplemented exception-throwing to fused batch-norm

### PR DESCRIPTION
**Pull request by duncanriach**
This PR adds determinism-unimplemented exception-throwing to `tf.compat.v1.nn.fused_batch_norm`. If an attempt is made to run the GPU kernel that implements backprop to `x`, `scale`, or `offset` when `is_training=False` (when fine-tuning) and when determinism is expected (i.e. when `TF_DETERMINISTIC_OPS` is set to `"true"` or `"1"`), then a `tf.errors.UnimplementedError` will be thrown.

This PR tests that the exception is thrown as appropriate and also tests that the other paths through the op (on both CPU and GPU) operate deterministically.

This PR is related to [RFC: Enabling Determinism in TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20210119-determinism.md). For status and history of GPU-determinism for this op, see [here](https://github.com/NVIDIA/framework-determinism/blob/master/tensorflow_status.md#fused-batch-norm).

CC @reedwm, @sanjoy, @nluehr

### CODE REVIEW

File: tensorflow/python/BUILD [hunk 1]
```diff
@@ -2917,6 +2917,28 @@ cuda_py_test(
     ],
 )
 
+cuda_py_test(
+    name = "nn_fused_batchnorm_deterministic_test",
+    size = "small",
+    srcs = ["ops/nn_fused_batchnorm_deterministic_test.py"],
+    python_version = "PY3",
+    shard_count = 1,
```
Comment (reedwm) [inline tensorflow/python/BUILD:2925]: Remove `shard_count = 1`, as 1 is the default

File: tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    variance = np.random.normal(size=scale_shape)
+    y_shape = x_shape; y_dtype = x_dtype
+    upstream_gradients = constant_op.constant(
+        np.random.normal(size=y_shape), dtype=y_dtype)
+    return x, scale, offset, mean, variance, upstream_gradients
+
+  @test_util.run_in_graph_and_eager_modes
+  def testForward(self):
+    with self.cached_session():
+      np.random.seed(123)
```
Comment (reedwm) [inline tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py:82]: The seed is [automatically set](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/framework/test_util.py;l=2162;drc=67c19451cedcc0774154a386da21f6c42ebbe73f) in the base class's setUp method, so setting here is unncessary.

File: tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                  tape.watch(backprop_to)
+                  op_output = nn_impl.fused_batch_norm(
+                      x, scale, offset, mean, variance, data_format=data_format,
+                      is_training=is_training, exponential_avg_factor=0.99)
+                  gradient_injector_output = op_output[0] * upstream_gradients
+                if (len(config.list_physical_devices('GPU')) > 0 and
+                    is_training == False):
+                  # Only backprop to offset is nondeterministic (on GPU, when
+                  # is_training=False), but backprop to the other parameters is
+                  # calculated using the same same kernel.
```
Comment (reedwm) [inline tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py:131]: The word "same" is duplicated

File: tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                  with self.assertRaisesRegex(
+                      errors_impl.UnimplementedError,
+                      "A deterministic GPU implementation of fused batch-norm" +
+                      " backprop, when training is disabled, is not currently" +
+                      " available."):
+                    grad = tape.gradient(gradient_injector_output, backprop_to)
+                    grad = self.evaluate(grad)
```
Comment (reedwm) [inline tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py:138]: No need to assign to grad. This line can be replaced with `self.evaluate(grad)`

File: tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            for is_training in [False, True]:
+              for backprop_to in [x, scale, offset]:
+                with backprop.GradientTape(persistent=True) as tape:
+                  tape.watch(backprop_to)
+                  op_output = nn_impl.fused_batch_norm(
+                      x, scale, offset, mean, variance, data_format=data_format,
+                      is_training=is_training, exponential_avg_factor=0.99)
+                  gradient_injector_output = op_output[0] * upstream_gradients
+                if (len(config.list_physical_devices('GPU')) > 0 and
+                    is_training == False):
```
Comment (reedwm) [inline tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py:128]: These two lines can be slightly simplified to

```python
                if (config.list_physical_devices('GPU') and
                    not is_training):
```

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py (hunk 2 above)
Comment (duncanriach) [reply tensorflow/python/ops/nn_fused_batchnorm_deterministic_test.py:82]: Wonderful. Thanks for revealing that.

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Review by Neilblaze — APPROVED** [review]
